### ISSUE TITLE
Not able to properly install/compile XGBoost for R with GPU support on RHEL

### ISSUE THREAD

**Issue by tsizzle** [dmlc/xgboost#5374]
Having great difficulty trying to install XGBoost for R with GPU support on RHEL 7.6.  I'm able run cmake 

`# cmake .. -DUSE_CUDA=ON -DR_LIB=ON -DCUDA_TOOLKIT_ROOT_DIR=/usr/local/cuda-10.1 -DCMAKE_C_COMPILER=cuda-gcc -DCMAKE_CXX_COMPILER=cuda-g++`
```
-- The CUDA compiler identification is NVIDIA 10.1.105
-- Check for working CUDA compiler: /usr/local/cuda-10.1/bin/nvcc
-- Check for working CUDA compiler: /usr/local/cuda-10.1/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Found OpenCV: /root/opencv-4.2.0/build (found version "4.2.0") 
-- Found CUDA: /usr/local/cuda-10.1 (found version "10.1") 
-- Configuring done
-- Generating done
```
```
CMake Warning:
  Manually-specified variables were not used by the project:

    CMAKE_CXX_COMPILER
    R_LIB


-- Build files have been written to: /opt/R-3.6.2/lib64/R/library/xgboost/build'
```

```
# make -j4
Scanning dependencies of target objxgboost
[100%] Building CUDA object CMakeFiles/objxgboost.dir/CMakeFiles/3.15.5/CompilerIdCUDA/CMakeCUDACompilerId.cu.o
[100%] Built target objxgboost
```

I'm unable to run `make install` as the Makefile that was created doesn't contain a "install:" target

```
# make install
make: *** No rule to make target `install'.  Stop.
```

But maybe this isn't necessary??? 

As a test, I tried running the demo  **`Rscript basic_walkthrough.R`** which worked fine.  But when I tried to run **`Rscript gpu_accelerated.R`** this failed and said that XGBoost version not compiled with GPU support.

```
# Rscript gpu_accelerated.R 
Error in xgb.iter.update(bst$handle, dtrain, iteration - 1, obj) : 
  [18:48:11] amalgamation/../src/learner.cc:180: XGBoost version not compiled with GPU support.

Stack trace returned 10 entries:
[bt] (0) /opt/R-3.6.2/lib64/R/library/xgboost/libs/xgboost.so(dmlc::StackTrace[abi:cxx11](unsigned long)+0x1ed) [0x7fac1a56c80d]
[bt] (1) /opt/R-3.6.2/lib64/R/library/xgboost/libs/xgboost.so(dmlc::LogMessageFatal::~LogMessageFatal()+0x1d) [0x7fac1a56d89d]
[bt] (2) /opt/R-3.6.2/lib64/R/library/xgboost/libs/xgboost.so(xgboost::LearnerImpl::AssertGPUSupport()+0x3c) [0x7fac1a59b18c]
[bt] (3) /opt/R-3.6.2/lib64/R/library/xgboost/libs/xgboost.so(xgboost::LearnerImpl::ConfigureUpdaters()+0x2a5) [0x7fac1a5e4e75]
[bt] (4) /opt/R-3.6.2/lib64/R/library/xgboost/libs/xgboost.so(xgboost::LearnerImpl::Configure(std::vector<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_trai
Calls: xgb.train -> xgb.iter.update
Execution halted
```

I must be doing something wrong. Thanks in advance for any help.

**Comment by hcho3** [comment]
Where did you obtain the source? I suggest that you use git to checkout the source.

git clone --recursive https://github.com/dmlc/xgboost

**Comment by tsizzle** [comment]
I got the tarball file from CRAN:  [https://cran.r-project.org/src/contrib/xgboost_0.90.0.2.tar.gz](https://cran.r-project.org/src/contrib/xgboost_0.90.0.2.tar.gz)  Is that ok to use?

**Comment by hcho3** [comment]
No, please use git if you’d like to enable GPU.

**Comment by tsizzle** [comment]
The environment I'm working is closed and does not have network access to GitHub.  I can only install via downloading the tarball file and/or install from source. I'm somewhat new to this.  Could I download the latest xgboost-master.zip and install from the /R-package folder?

**Comment by trivialfis** [comment]
No, you can't.  The zip file provided by github doesn't include submodules.  I'm confused, if you don't have network access how do you download tar ball?

**Comment by tsizzle** [comment]
The RHEL target machine is completely closed off (internal network only).  I have to download the file onto a usb flash drive and manually copy it over.  Yes, it feels like the 1990s.  So are you saying the 0.90.02 tar ball file from CRAN is not usable at all for installing and compiling xgboost to work with GPU?

**Comment by hcho3** [comment]
You can use git to obtain the source code and create a tarball.

tar cvjf xgboost.tar.bz2 xgboost/

**Comment by trivialfis** [comment]
I suggest cloning the 1.0.0 release.  There's a bug in GPU #5351 on master branch, but not in 1.0 release branch.

**Comment by hcho3** [comment]
> the 0.90.02 tar ball file from CRAN is not usable at all for installing and compiling xgboost to work with GPU?

That’s correct. The source tarball uploaded to CRAN only contains CPU code. The reason is that CRAN requires support for multiple platforms, many of which does not support CUDA.

**Comment by tsizzle** [comment]
Ok, thanks, I'll give it a try with the 1.0.0 release.  Once the tarball is created and copied to the RHEL target machine, I can run R and `install.packages([R-package src], repo=NULL, type="source")`?  Might you know where I can copy the proper **CMakeLists.txt** file for running `cmake .. -DUSE_CUDA=ON -DR_LIB=ON ` ?  Apologies in advance for being new to this.

**Comment by hcho3** [comment]
CMakeLists.txt is located at the root of XGBoost git repo. Run
```
tar xvjf xgboost.tar.bz2
cd xgboost
mkdir build
cd build
cmake .. -DR_LIB=ON -DUSE_CUDA=ON
make install
```

Do not use `install.packages`

**Comment by tsizzle** [comment]
I'm running into this error during cmake:
```
CMake Error at /usr/local/share/cmake-3.15/Modules/FindPackageHandleStandardArgs.cmake:137 (message):
  Could NOT find LibR (missing: LIBR_CORE_LIBRARY)
Call Stack (most recent call first):
  /usr/local/share/cmake-3.15/Modules/FindPackageHandleStandardArgs.cmake:378 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindLibR.cmake:173 (find_package_handle_standard_args)
  R-package/CMakeLists.txt:1 (find_package)
```

Is it trying to look for my LIB_CORE_LIBRARY path?  What path should that be?

**Comment by tsizzle** [comment]
Finally got cmake to run after tweaking the R-package/CMakeList.txt with `set(LIBR_CORE_LIBRARY /usr/R.3.6.2/lib64)`  had a bunch of warnings.  However `make install` failed.

```
[root@localhost xgboost]# rm -rf build
[root@localhost xgboost]# mkdir build
[root@localhost xgboost]# cd build
[root@localhost build]# cmake .. -DUSE_CUDA=ON -DR_LIB=ON
-- The CXX compiler identification is GNU 6.5.0
-- The C compiler identification is GNU 6.5.0
-- Check for working CXX compiler: /usr/local/bin/c++
-- Check for working CXX compiler: /usr/local/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Check for working C compiler: /usr/local/bin/gcc
-- Check for working C compiler: /usr/local/bin/gcc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- CMake version 3.15.5
-- Performing Test XGBOOST_MM_PREFETCH_PRESENT
-- Performing Test XGBOOST_MM_PREFETCH_PRESENT - Success
-- Performing Test XGBOOST_BUILTIN_PREFETCH_PRESENT
-- Performing Test XGBOOST_BUILTIN_PREFETCH_PRESENT - Success
-- xgboost VERSION: 1.0.0
-- Setting build type to 'Release' as none was specified.
-- Configured CUDA host compiler: /usr/local/bin/c++
-- The CUDA compiler identification is NVIDIA 10.1.105
-- Check for working CUDA compiler: /usr/local/cuda-10.1/bin/nvcc
-- Check for working CUDA compiler: /usr/local/cuda-10.1/bin/nvcc -- works
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- CUDA GEN_CODE: --generate-code=arch=compute_35,code=sm_35;--generate-code=arch=compute_50,code=sm_50;--generate-code=arch=compute_52,code=sm_52;--generate-code=arch=compute_60,code=sm_60;--generate-code=arch=compute_61,code=sm_61;--generate-code=arch=compute_70,code=sm_70;--generate-code=arch=compute_75,code=sm_75;--generate-code=arch=compute_75,code=compute_75;
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Found OpenMP: TRUE (found version "4.5")  
-- Performing Test SUPPORT_CXX11
-- Performing Test SUPPORT_CXX11 - Success
-- Performing Test SUPPORT_CXX0X
-- Performing Test SUPPORT_CXX0X - Success
-- Performing Test SUPPORT_MSSE2
-- Performing Test SUPPORT_MSSE2 - Success
-- Found OpenMP_C: -fopenmp (found version "4.5") 
-- Found OpenMP_CXX: -fopenmp (found version "4.5") 
-- Could NOT find GTest (missing: GTEST_LIBRARY GTEST_INCLUDE_DIR GTEST_MAIN_LIBRARY) 
CMake Warning at dmlc-core/test/unittest/CMakeLists.txt:37 (message):
  Google Test not found


-- Found LibR: /opt/R-3.6.2/lib64/R  
-- Found R: /opt/R-3.6.2/bin/R
-- LIBR_CORE_LIBRARY /opt/R-3.6.2/lib64
-- Configuring done
WARNING: Target "runxgboost" requests linking to directory "/opt/R-3.6.2/lib64".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "xgboost" requests linking to directory "/opt/R-3.6.2/lib64".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "xgboost" requests linking to directory "/opt/R-3.6.2/lib64".  Targets may link only to libraries.  CMake is dropping the item.
WARNING: Target "runxgboost" requests linking to directory "/opt/R-3.6.2/lib64".  Targets may link only to libraries.  CMake is dropping the item.
-- Generating done
-- Build files have been written to: /opt/R-3.6.2/lib64/R/library/xgboost/build
```

Then when I run `make install `:

```
[root@localhost build]# make install
Scanning dependencies of target objxgboost
[  1%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api.cc.o
[  2%] Building CXX object src/CMakeFiles/objxgboost.dir/c_api/c_api_error.cc.o
[  3%] Building CXX object src/CMakeFiles/objxgboost.dir/common/common.cc.o
In file included from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/common.cc:10:0:
/opt/R-3.6.2/lib64/R/library/xgboost/src/common/./random.h: In constructor â€˜xgboost::common::ColumnSampler::ColumnSampler()â€™:
/opt/R-3.6.2/lib64/R/library/xgboost/src/common/./random.h:130:52: error: no matching function for call to â€˜Broadcast(uint32_t*, long unsigned int, int, const char [5])â€™
     rabit::Broadcast(&seed, sizeof(seed), 0, "seed");
                                                    ^
In file included from /opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/rabit.h:354:0,
                 from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/./random.h:10,
                 from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/common.cc:10:
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:130:13: note: candidate: void rabit::Broadcast(void*, size_t, int)
 inline void Broadcast(void *sendrecv_data, size_t size, int root) {
             ^~~~~~~~~
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:130:13: note:   candidate expects 3 arguments, 4 provided
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:134:13: note: candidate: template<class DType> void rabit::Broadcast(std::vector<T>*, int)
 inline void Broadcast(std::vector<DType> *sendrecv_data, int root) {
             ^~~~~~~~~
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:134:13: note:   template argument deduction/substitution failed:
In file included from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/common.cc:10:0:
/opt/R-3.6.2/lib64/R/library/xgboost/src/common/./random.h:130:52: note:   mismatched types â€˜std::vector<T>â€™ and â€˜uint32_t {aka unsigned int}â€™
     rabit::Broadcast(&seed, sizeof(seed), 0, "seed");
                                                    ^
In file included from /opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/rabit.h:354:0,
                 from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/./random.h:10,
                 from /opt/R-3.6.2/lib64/R/library/xgboost/src/common/common.cc:10:
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:144:13: note: candidate: void rabit::Broadcast(std::__cxx11::string*, int)
 inline void Broadcast(std::string *sendrecv_data, int root) {
             ^~~~~~~~~
/opt/R-3.6.2/lib64/R/library/xgboost/rabit/include/rabit/./internal/rabit-inl.h:144:13: note:   candidate expects 2 arguments, 4 provided
make[2]: *** [src/CMakeFiles/objxgboost.dir/common/common.cc.o] Error 1
make[1]: *** [src/CMakeFiles/objxgboost.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by hcho3** [comment]
Did you check out XGBoost repo and then later switch to release_1.0.0 branch? It looks like you have a wrong version of Rabit module. Try running `git submodule update - -init - -recursive`
